plutus-tx/src/PlutusTx/Foldable.hs

```diff
@@ -37,75 +38,74 @@ module PlutusTx.Foldable (
   ) where
 
 import Control.Applicative (Alternative (..), Const (..))
-import Data.Coerce (Coercible, coerce)
 import Data.Functor.Identity (Identity (..))
-import Data.Monoid (First (..))
-import Data.Semigroup (Dual (..), Endo (..), Product (..), Sum (..))
 import GHC.Exts (build)
 import PlutusTx.Applicative (Applicative (pure), (*>))
 import PlutusTx.Base
-import PlutusTx.Bool (Bool (..), not)
+import PlutusTx.Bool (Bool (..), not, (&&), (||))
 import PlutusTx.Builtins (Integer)
 import PlutusTx.Either (Either (..))
 import PlutusTx.Eq (Eq (..))
 import PlutusTx.Maybe (Maybe (..))
 import PlutusTx.Monoid (Monoid (..))
-import PlutusTx.Numeric (AdditiveMonoid, AdditiveSemigroup ((+)), MultiplicativeMonoid)
+import PlutusTx.Numeric
 import PlutusTx.Semigroup ((<>))
 
 import Prelude qualified as Haskell (Monad, return, (>>), (>>=))
 
 -- | Plutus Tx version of 'Data.Foldable.Foldable'.
 class Foldable t where
-    -- | Plutus Tx version of 'Data.Foldable.foldMap'.
-    foldMap :: Monoid m => (a -> m) -> t a -> m
+    -- | Plutus Tx version of 'Data.Foldable.foldr'.
+    foldr :: (a -> b -> b) -> b -> t a -> b
 
     -- All the other methods are deliberately omitted,
     -- to make this a one-method class which has a simpler representation
 
 instance Foldable [] where
-    {-# INLINABLE foldMap #-}
-    foldMap _ []     = mempty
-    foldMap f (x:xs) = f x <> foldMap f xs
+    {-# INLINABLE foldr #-}
+    foldr f z = go
+      where
+        go = \case
+          []     -> z
+          x : xs -> f x (go xs)
```

Comment:
What about `oneShot (f x) (go xs)` then? I know we probably shouldn't try to do fancy things to get efficient compilation, because it's way too unreliable, but maybe a few general policies here and there would be fine?